x/pki/handler_add_non_root_cert_test.go

```diff
@@ -0,0 +1,636 @@
+	_, err := setup.Handler(setup.Ctx, addNocX509RootCert)
+	require.NoError(t, err)
+
+	// add x509 certificate
+	addX509Cert := types.NewMsgAddX509Cert(accAddress.String(), testconstants.IntermediateCertPem)
+	_, err = setup.Handler(setup.Ctx, addX509Cert)
+	require.ErrorIs(t, err, pkitypes.ErrInappropriateCertificateType)
+}
+
+func TestHandler_AddX509Cert_ForDifferentSerialNumber(t *testing.T) {
+	setup := Setup(t)
+
+	// store root certificate
+	rootCertificate := rootCertificate(setup.Trustee1)
+	setup.Keeper.AddApprovedCertificate(setup.Ctx, rootCertificate)
+
+	vendorAccAddress := GenerateAccAddress()
+	setup.AddAccount(vendorAccAddress, []dclauthtypes.AccountRole{dclauthtypes.Vendor}, testconstants.Vid)
+
+	// store intermediate certificate with different serial number
+	intermediateCertificate := intermediateCertificate(vendorAccAddress)
+	intermediateCertificate.SerialNumber = SerialNumber
+	setup.Keeper.SetUniqueCertificate(
+		setup.Ctx,
+		uniqueCertificate(intermediateCertificate.Issuer, intermediateCertificate.SerialNumber),
+	)
+	setup.Keeper.AddApprovedCertificate(setup.Ctx, intermediateCertificate)
+
+	// store intermediate certificate second time
+	addX509Cert := types.NewMsgAddX509Cert(vendorAccAddress.String(), testconstants.IntermediateCertPem)
+	_, err := setup.Handler(setup.Ctx, addX509Cert)
+	require.NoError(t, err)
+
+	// query certificate
+	certificates, _ := queryApprovedCertificates(setup, testconstants.IntermediateSubject, testconstants.IntermediateSubjectKeyID)
+
+	// check
+	require.Equal(t, 2, len(certificates.Certs))
+	require.NotEqual(t, certificates.Certs[0].SerialNumber, certificates.Certs[1].SerialNumber)
+
+	for _, certificate := range certificates.Certs {
+		require.Equal(t, addX509Cert.Cert, certificate.PemCert)
+		require.Equal(t, addX509Cert.Signer, certificate.Owner)
+		require.Equal(t, testconstants.IntermediateSubject, certificate.Subject)
+		require.Equal(t, testconstants.IntermediateSubjectKeyID, certificate.SubjectKeyId)
+		require.False(t, certificate.IsRoot)
+		require.Equal(t, testconstants.RootSubject, certificate.RootSubject)
+		require.Equal(t, testconstants.RootSubjectKeyID, certificate.RootSubjectKeyId)
+		require.Equal(t, testconstants.IntermediateIssuer, certificate.Issuer)
+		require.Equal(t, testconstants.IntermediateAuthorityKeyID, certificate.AuthorityKeyId)
+	}
+}
+
+func TestHandler_AddX509Cert_ForAbsentDirectParentCert(t *testing.T) {
+	setup := Setup(t)
+
+	vendorAccAddress := GenerateAccAddress()
+	setup.AddAccount(vendorAccAddress, []dclauthtypes.AccountRole{dclauthtypes.Vendor}, testconstants.Vid)
+
+	// add intermediate x509 certificate
+	addX509Cert := types.NewMsgAddX509Cert(vendorAccAddress.String(), testconstants.IntermediateCertPem)
+	_, err := setup.Handler(setup.Ctx, addX509Cert)
+	require.ErrorIs(t, err, pkitypes.ErrInvalidCertificate)
+}
+
+func TestHandler_AddX509Cert_ForFailedCertificateVerification(t *testing.T) {
```

Comment:
```suggestion
func TestHandler_AddX509Cert_WhenCertificateVerificationFails(t *testing.T) {
```

Reply:
Let's keep it for consistency with other test cases names